### PR TITLE
Feature/search library

### PR DESCRIPTION
**Pull request by vitalie-ucommerce**
Introduce the new `SearchLibrary`.
Fix `SearchController`, `CategoryController`, `FacetsController`.

### CODE REVIEW

File: src/Ucommerce.Avenue.Umbraco/Controllers/CategoryController.cs [hunk 1]
```diff
@@ -1,49 +1,46 @@
...
 using Umbraco.Web.Models;
 
 namespace UCommerce.RazorStore.Controllers
 {
     public class CategoryController : RenderMvcController
     {
+        private readonly ICatalogLibrary _catalogLibrary = ObjectFactory.Instance.Resolve<ICatalogLibrary>();
+        private readonly ISiteContext _siteContext = ObjectFactory.Instance.Resolve<ISiteContext>();
+        private readonly ISearchLibrary _searchLibrary = ObjectFactory.Instance.Resolve<ISearchLibrary>();
+
```
Comment (ilny) [inline src/Ucommerce.Avenue.Umbraco/Controllers/CategoryController.cs]: @skovsboll Should we do
A) private readonly ICatalogContext _catalogContext
B) public ICatalogContext CatalogContext
?

Let's pick one and be consistent with our naming :)

File: src/Ucommerce.Avenue.Umbraco/Controllers/CategoryController.cs [hunk 2]
```diff
@@ -1,49 +1,46 @@
...
+            var currentCategory = _siteContext.CatalogContext.CurrentCategory;
 
             var categoryViewModel = new CategoryViewModel
             {
-                Name = currentCategory.DisplayName(),
-                Description = currentCategory.DisplayName(),
-                CatalogId = currentCategory.ProductCatalog.Id,
-                CategoryId = currentCategory.Id,
+                Name = currentCategory.DisplayName,
+                Description = currentCategory.DisplayName,
```
Comment (ilny) [inline src/Ucommerce.Avenue.Umbraco/Controllers/CategoryController.cs]: Make it `Description = currentCategory.Description` instead, the same nonsense was present in **ProductController as well**.

File: src/Ucommerce.Avenue.Umbraco/Controllers/SearchController.cs [hunk 3]
```diff
@@ -1,64 +1,63 @@
...
-                    LongDescription = product.LongDescription(),
-                    // PriceCalculation = CatalogLibrary.CalculatePrices(new List<Guid>{product.Guid}),
-                    ThumbnailImageUrl = ObjectFactory.Instance.Resolve<IImageService>()
-                        .GetImage(product.ThumbnailImageMediaId).Url,
+                    IsVariant = !string.IsNullOrWhiteSpace(product.VariantSku),
+                    LongDescription = product.LongDescription,
+                    PriceCalculation = new ProductPriceCalculationViewModel
+                    {
+                        YourPrice = productPriceCalculationResultItem.PriceInclTax.ToString("C"),
+                        ListPrice = productPriceCalculationResultItem.ListPriceInclTax.ToString("C")
```
Comment (ilny) [inline src/Ucommerce.Avenue.Umbraco/Controllers/SearchController.cs]: What is the reason of passing string `C`?

File: src/Ucommerce.Avenue.Umbraco/Controllers/CategoryController.cs (hunk 1 above)
Comment (vitalie-ucommerce) [reply src/Ucommerce.Avenue.Umbraco/Controllers/CategoryController.cs]: I guess we go with A.
Good catch

File: src/Ucommerce.Avenue.Umbraco/Controllers/CategoryController.cs (hunk 2 above)
Comment (vitalie-ucommerce) [reply src/Ucommerce.Avenue.Umbraco/Controllers/CategoryController.cs]: yep, changed it

File: src/Ucommerce.Avenue.Umbraco/Controllers/SearchController.cs (hunk 3 above)
Comment (vitalie-ucommerce) [reply src/Ucommerce.Avenue.Umbraco/Controllers/SearchController.cs]: so the decimal is displayed 'pretty'. With this it's like this: `$320.85`, without it something like this: `320.85000`

File: src/Ucommerce.Avenue.Umbraco/Api/AvenueClothingApiProductController.cs [hunk 4]
```diff
@@ -4,63 +4,61 @@
...
+        public ICatalogLibrary CatalogLibrary => ObjectFactory.Instance.Resolve<ICatalogLibrary>();
+        public ICatalogContext CatalogContext => ObjectFactory.Instance.Resolve<ICatalogContext>();
+        public IIndex<Product> ProductsIndex => ObjectFactory.Instance.Resolve<IIndex<Product>>();
 
         [Route("razorstore/products/getproductvariations")]
         [HttpPost]
         public IHttpActionResult GetProductVariations([FromBody] GetProductVariationsRequest request)
         {
-            var product = _catalogLibrary.GetProduct(request.ProductSku);
+            Search.Models.Product product =
```
Comment (ilny) [inline src/Ucommerce.Avenue.Umbraco/Api/AvenueClothingApiProductController.cs]: Why the need to specify the namespace if we already have an alias on line 13?
Is there a chance to completely remove the `using` of  `EntitiesV2` (and thus the aliases as well) at this point?

Comment (vitalie-ucommerce) [reply src/Ucommerce.Avenue.Umbraco/Api/AvenueClothingApiProductController.cs]: we don't need to specify namespaces, you are right.
But we can't remove EnititesV2 yet, we get the Currency from there (line 91)

File: src/Ucommerce.Avenue.Umbraco/Api/AvenueClothingApiSearchController.cs [hunk 5]
```diff
@@ -6,35 +6,32 @@
...
+            var products = ProductIndex.Find()
+                .Where(p => p.VariantSku == null
+                            && (
+                                p.Sku.Contains(request.Keyword)
+                                || p.Name.Contains(request.Keyword)
+                                || p.DisplayName.Contains(request.Keyword)
+                                || p.LongDescription.Contains(request.Keyword)
+                                || p.ShortDescription.Contains(request.Keyword)
+                            ))
+                .ToList();
```
Comment (skovsboll) [inline src/Ucommerce.Avenue.Umbraco/Api/AvenueClothingApiSearchController.cs]: It's great that the query translates so 1:1 from Linq to Bolt! 😉 
However, shouldn't we rather use `p.DisplayName == Match.Fulltext(request.Keyword)` in order to use the fact that we full text index some of the fields?

File: src/Ucommerce.Avenue.Umbraco/Api/AvenueClothingApiSearchController.cs [hunk 6]
```diff
@@ -53,15 +50,8 @@ private IList<ProductVariation> MapResult(IList<UCommerce.EntitiesV2.Product> pr
...
-                    Properties = product.ProductProperties.Select(prop => new ProductProperty()
-                    {
-                        Id = prop.Id,
-                        Name = prop.ProductDefinitionField.Name,
-                        Value = prop.Value
-                    }),
-                    // TODO: re-implement
-                    // Url = UrlService.GetUrl(CatalogContext.CurrentCatalog, new[] {product})
+                    ProductName = product.DisplayName,
+                    Url = UrlService.GetUrl(CatalogContext.CurrentCatalog, new[] { product })
```
Comment (skovsboll) [inline src/Ucommerce.Avenue.Umbraco/Api/AvenueClothingApiSearchController.cs]: Shorter. Nicer.
:100:

File: src/Ucommerce.Avenue.Umbraco/Controllers/CategoryController.cs [hunk 7]
```diff
@@ -68,12 +63,13 @@ private IList<ProductViewModel> MapProductsInCategories(Category category)
             IList<Facet> facetsForQuerying = System.Web.HttpContext.Current.Request.QueryString.ToFacets();
             var productsInCategory = new List<ProductViewModel>();
 
-            foreach (var subcategory in category.Categories)
+            foreach (var subCategoryGuid in category.Categories)
             {
-                productsInCategory.AddRange(MapProductsInCategories(subcategory));
+                var subCategory = CatalogLibrary.GetCategory(subCategoryGuid);
```
Comment (skovsboll) [inline src/Ucommerce.Avenue.Umbraco/Controllers/CategoryController.cs]: There's a _SELECT N+1_ here. Not going to perform well.
Better option would be to get all sub cats in one query, to a `.ToLookup` and then index that lookup inside the loop.

The fact that it's recursive makes it hard to make it perform well. For the category navigation component, @mufoxe and I agreed that for now, we'd support two levels and not make it recursive. Later, we'll add js code to load more levels. I'm not sure if the same logic is applicable here, though?

File: src/Ucommerce.Avenue.Umbraco/Controllers/SearchController.cs (hunk 3 above)
Comment (skovsboll) [reply src/Ucommerce.Avenue.Umbraco/Controllers/SearchController.cs]: Yes, but we need to redo this part. There's a bunch of specialized code for formatting money. ToString is too simplistic (it doesn't look up the currency symbol in based on the pricegroup for instance).

Comment (vitalie-ucommerce) [reply src/Ucommerce.Avenue.Umbraco/Controllers/SearchController.cs]: yes of course, but I think we should tackle that after the workshop

File: src/Ucommerce.Avenue.Umbraco/Api/AvenueClothingApiSearchController.cs (hunk 5 above)
Comment (vitalie-ucommerce) [reply src/Ucommerce.Avenue.Umbraco/Api/AvenueClothingApiSearchController.cs]: Yep, my bad.

File: src/Ucommerce.Avenue.Umbraco/Controllers/CategoryController.cs (hunk 7 above)
Comment (vitalie-ucommerce) [reply src/Ucommerce.Avenue.Umbraco/Controllers/CategoryController.cs]: you are right.
I was too focused at replicating all 1:1, so I didn't think to use what the new structure to our advantage

**Comment by skovsboll** [comment]
For those of you reviewing this PR, you need to manually delete the Apps/RavenDB25 folder before it works.

File: src/Ucommerce.Avenue.Umbraco/Controllers/SearchController.cs (hunk 3 above)
Comment (skovsboll) [reply src/Ucommerce.Avenue.Umbraco/Controllers/SearchController.cs]: :+1:

File: src/Ucommerce.Avenue.Umbraco/Api/AvenueClothingApiSearchController.cs [hunk 8]
```diff
@@ -27,7 +27,7 @@ public IHttpActionResult Search([FromBody] SearchRequest request)
                             && (
                                 p.Sku.Contains(request.Keyword)
                                 || p.Name.Contains(request.Keyword)
-                                || p.DisplayName.Contains(request.Keyword)
+                                || p.DisplayName == Match.FullText(request.Keyword)
                                 || p.LongDescription.Contains(request.Keyword)
                                 || p.ShortDescription.Contains(request.Keyword)
```
Comment (skovsboll) [inline src/Ucommerce.Avenue.Umbraco/Api/AvenueClothingApiSearchController.cs]: Long and Short description are also full test indexed: https://github.com/Ucommercenet/Ucommerce/blob/16713971f2d15a48e108a54a7decaa4fbd937365/src/Ucommerce.Search/Definitions/DefaultProductsIndexDefinition.cs#L33

File: src/Ucommerce.Avenue.Umbraco/Controllers/CategoryController.cs [hunk 9]
```diff
@@ -63,10 +63,11 @@ private IList<ProductViewModel> MapProductsInCategories(Category category)
...
             var productsInCategory = new List<ProductViewModel>();
 
-            foreach (var subCategoryGuid in category.Categories)
+            var subCategories = CatalogLibrary.GetCategories(category.Categories);
+
+            foreach (var subCategory in subCategories)
             {
-                var subCategory = CatalogLibrary.GetCategory(subCategoryGuid);
-                productsInCategory.AddRange(MapProductsInCategories(subCategory));
+                productsInCategory.AddRange(MapProducts(SearchLibrary.GetProductsFor(subCategory.Guid, facetsForQuerying).Results));
```
Comment (skovsboll) [inline src/Ucommerce.Avenue.Umbraco/Controllers/CategoryController.cs]: This is still, technically, a SELECT N+1.
All retrieval should be outside the `foreach`.

File: src/Ucommerce.Avenue.Umbraco/Controllers/CategoryController.cs (hunk 7 above)
Comment (skovsboll) [reply src/Ucommerce.Avenue.Umbraco/Controllers/CategoryController.cs]: There's still a SELECT N+1
If you have a query inside a loop, that's one.
All querying should be done before the `foreach`.

File: src/Ucommerce.Avenue.Umbraco/Controllers/CategoryController.cs (hunk 9 above)
Comment (skovsboll) [reply src/Ucommerce.Avenue.Umbraco/Controllers/CategoryController.cs]: Let's get it in, and I'll help fix this later.

**Pull request merged**